File: workers/go/harness/profiles.go
package harness

import (
	"fmt"

	"go.temporal.io/sdk/contrib/sysinfo"
	sdkworker "go.temporal.io/sdk/worker"
)

const (
	workerProfileEnvVar                   = "OMES_WORKER_PROFILE"
	resourceBasedDefaultWorkerProfile     = "resource-based-default"
	throughputStressBaselineWorkerProfile = "throughput-stress-baseline"
)

type workerProfile struct {
	Options                 sdkworker.Options
	StickyWorkflowCacheSize int
}

func registerWorkerProfile(name string, profile workerProfile) {
	workerProfiles[name] = profile
}

func lookupWorkerProfile(name string) (workerProfile, error) {
	profile, ok := workerProfiles[name]
	if !ok {
		return workerProfile{}, fmt.Errorf("unknown worker profile %q", name)
	}
	return profile, nil
}

var workerProfiles = map[string]workerProfile{}

func init() {
	registerWorkerProfile(resourceBasedDefaultWorkerProfile, workerProfile{
		Options: sdkworker.Options{
			Tuner: mustResourceBasedTuner(0.8, 0.8),
		},
	})
	registerWorkerProfile(throughputStressBaselineWorkerProfile, workerProfile{
		StickyWorkflowCacheSize: 50,
		Options: sdkworker.Options{
			MaxConcurrentWorkflowTaskExecutionSize:  8,
			MaxConcurrentActivityExecutionSize:      32,
			MaxConcurrentLocalActivityExecutionSize: 32,
			MaxConcurrentWorkflowTaskPollers:        2,
			MaxConcurrentActivityTaskPollers:        4,
		},
	})
}

func mustResourceBasedTuner(targetMem, targetCpu float64) sdkworker.WorkerTuner {
	tuner, err := sdkworker.NewResourceBasedTuner(sdkworker.ResourceBasedTunerOptions{
		TargetMem:    targetMem,
		TargetCpu:    targetCpu,
		InfoSupplier: sysinfo.SysInfoProvider(),
	})
	if err != nil {
		panic(err)
	}
	return tuner
}

File: workers/ruby/harness/profiles.rb
# frozen_string_literal: true

require 'temporalio/worker'

module Harness
  module Profiles
    WORKER_PROFILE_ENV_VAR = 'OMES_WORKER_PROFILE'
    RESOURCE_BASED_DEFAULT_PROFILE = 'resource-based-default'
    THROUGHPUT_STRESS_BASELINE_PROFILE = 'throughput-stress-baseline'

    @registry = {}

    class << self
      def register(name, profile)
        @registry[name] = profile
      end

      def lookup(name)
        @registry.fetch(name).dup
      rescue KeyError
        raise ArgumentError, "Unknown worker profile #{name.inspect}"
      end
    end

    register(
      RESOURCE_BASED_DEFAULT_PROFILE,
      {
        tuner: Temporalio::Worker::Tuner.create_resource_based(
          target_memory_usage: 0.8,
          target_cpu_usage: 0.8
        )
      }
    )

    register(
      THROUGHPUT_STRESS_BASELINE_PROFILE,
      {
        tuner: Temporalio::Worker::Tuner.create_fixed(
          workflow_slots: 8,
          activity_slots: 32,
          local_activity_slots: 32
        ),
        max_cached_workflows: 50,
        max_concurrent_workflow_task_polls: 2,
        max_concurrent_activity_task_polls: 4
      }
    )
  end
end
